Bokeh -> Plotly.js: frontend-side plotting

Theodlz opened the pull request:
### This is a work in progress. Do not merge.

This PR is an attempt to reproduce the photometry and spectra plot using Plotly instead of Bokeh, to reduce latency significantly, and reduce the load put on the server side.

We'll need to play around with different ways to generate and update the plots, to see what works best performance-wise. A lot is already implemented, but some key features still need to be implemented.

I can already see a CLEAR boost in terms of speed, and significantly reduced bandwidth when going to a source page. However, since its all done on the frontend now, we might see some slowdowns when resizing and so on, to verify.

Identified TODOs left:
- [x] Redo the instrument summary plot (hard)
- [x] add the secondary X and Y axis on all the plots that have those.
- [x] smoothing of period plot and spectra (hard)
- [x] add the one phase / two phase button on the period plot (easy to medium)
- [x] map the sncosmo filters to associated colors (easy, mostly an implementation choice)
- [x] download button on the phot plot (easy)
- [x] add the button to create period annotations using the period plot (easy)
- [x] add the custom wavelength box on the spectra plot (easy)
- [x] probably fix some unit tests (easy)
- [x] cleanup of the code (very easy)

and in general, make it look nice!

Theodlz commented:
Ah and @profjsb I know you mentioned licensing issues with plotly in the past. I checked and it seems that now its a pretty classic and very permissive MIT license, which should be fine for us correct? Just want to make sure before I keep experimenting further with plotly.

mcoughlin reviewed (commented):
I notice the messages:

[17:37:26 api/photometry] f1500w with effective wavelength 150913.772228051 is out of range for color maps, using black
[17:37:26 api/photometry] f1800w with effective wavelength 180062.13892048062 is out of range for color maps, using black
[17:37:26 api/photometry] f2100w with effective wavelength 208424.90993173423 is out of range for color maps, using black
[17:37:27 api/photometry] f2550w with effective wavelength 254082.59789131142 is out of range for color maps, using black
[17:37:27 api/photometry] f1065c with effective wavelength 106499.99999999997 is out of range for color maps, using black
[17:37:27 api/photometry] f1140c with effective wavelength 114000.0 is out of range for color maps, using black

in the logs.

On the plots:

Everything looks nice. Do we not show the upperlimits by design in the period folding? It's fine, just never noticed it.

Theodlz commented:
> I notice the messages:
> 
> [17:37:26 api/photometry] f1500w with effective wavelength 150913.772228051 is out of range for color maps, using black [17:37:26 api/photometry] f1800w with effective wavelength 180062.13892048062 is out of range for color maps, using black [17:37:26 api/photometry] f2100w with effective wavelength 208424.90993173423 is out of range for color maps, using black [17:37:27 api/photometry] f2550w with effective wavelength 254082.59789131142 is out of range for color maps, using black [17:37:27 api/photometry] f1065c with effective wavelength 106499.99999999997 is out of range for color maps, using black [17:37:27 api/photometry] f1140c with effective wavelength 114000.0 is out of range for color maps, using black
> 
> in the logs.
> 
> On the plots:
> 
> Everything looks nice. Do we not show the upperlimits by design in the period folding? It's fine, just never noticed it.

Thanks for the feedback! 
1. About the warnings: this is a problem we already had, but never showed up as we never have any JWST data, but the color() method in plot.py doesn't handle their effective wavelength. I added a TODO where it needs to be added. Any suggestion as to what color I should use for the jwst-miri and jwst-miri-tophat bandpasses?
2. Good point. I had a doubt myself when re-implementing it. I'll double check and I'll make sure that it does the same thing as the current plot.

Any changes you think should be added to this PR?

mcoughlin commented:
@Theodlz not really for colors. The deepest red you can find. But to me, this all looks great.

Theodlz commented:
NOTE for the reviewers, and to keep track of that in general: When using react-plotly's default Plot, the webpack is VERY large. Instead, we import a minimized version of Plotly, and use react-plotly's factory to use that version instead of the full version. I found that here: https://github.com/plotly/react-plotly.js/issues/98, and it reduces the chuk size of plotly in the webpack by a factor of ~4 when gzipped.

Theodlz commented:
@guynir42 this is the PR mentioned during today's call. There's a checklist with the things that are missing before this can be merged.

guynir42 reviewed (approved):
The code looks good, as far as I can tell. It is sad to see all the bokeh stuff we invested so much time into just go away, but also I'm glad we are moving to something lighter and quicker (and maybe less complicated).

guynir42 commented:
It looks like the spectrum binning is doing the right thing. It doesn't show the original spectrum faded in the background like the old plot. Is that still an issue?

mcoughlin commented on package.json:
```diff
@@ -60,6 +57,7 @@
     "numeral": "2.0.6",
     "papaparse": "5.4.1",
     "path-browserify": "1.0.1",
+    "plotly.js-basic-dist": "^2.25.2",
```
Can you please remove ^ here and below?

mcoughlin commented on skyportal/handlers/api/photometry.py:
```diff
@@ -54,6 +56,80 @@
+
+def get_effective_wavelength(bandpass_name):
+    try:
+        bandpass = snc.get_bandpass(bandpass_name)
```
Make an optional argument for radial dependent bandpass a la: sncosmo.get_bandpass('"ultrasat', 5.0)

mcoughlin replied:
Ah some bands require a second argument that is numerical corresponding to position on the focal plane, like Ultrasat.

Theodlz replied:
Got it. So I can add the argument here without any problem, but I suppose we won't use it here for the plotting right?

mcoughlin replied:
Yeah not yet, just noticed we will need it for certain bands

mcoughlin commented on skyportal/tests/frontend/test_bokeh_plots_frontend.py:
```diff
@@ -16,22 +16,23 @@ def expand_shadow_element(driver, element):
 def test_export_bold_light_curve_as_csv_button(driver, user, public_source):
     driver.get(f"/become_user/{user.id}")
     driver.get(f"/source/{public_source.id}")
-    driver.wait_for_xpath('//div[@class=" bk-root"]', timeout=20)
```
Change name of this to test_plotly_plots_frontend?

Theodlz replied:
good idea. We might want to still test the export to CSV but via the photometry table

mcoughlin commented on static/js/components/InstrumentSummary.jsx:
```diff
@@ -60,9 +61,7 @@ const InstrumentPlot = ({ instrumentId, startDate, endDate }) => {
-      <Plot
-        url={`/api/internal/plot/instrument_log/${instrumentId}?width=${plotWidth}&height=500&startDate=${startDate}&endDate=${endDate}`}
-      />
+      <div>TODO</div>
```
Just flagging this here.

Theodlz replied:
yikes I forgot about that instruments log plot... will re-implement it too!

mcoughlin reviewed (approved):
Just a few things to note. Code is very nicely readable.

stefanv reviewed (approved):
Hi Theo, thanks for this enormous refactoring!

I looked at the code, and left a few comments here and there. Overall, it is very readable. It'd be good to add a few tests for utility functions, and comments where tricky things are done, but I realize that we don't have JS testing infrastructure in place.

The one place I think you may want to get an additional set of eyes is by a user of the platform, so they can check whether anything is missing / changing in behavior—I cannot verify that.

I noted that `smooth_spectra`, `foldphase` etc. are now gone, and it wasn't immediately clear to me that all of it is replaced (although I suspect it is, based on yesterday's demo).

My inclination is to get this merged sooner rather than later, and to then iterate on getting any wrinkles sorted out.

So, I recommend @mcoughlin or @profjsb play around with the PR locally, and if it looks good to them, go ahead and merge.

stefanv commented:
It may also be worth seeing whether a generic Plotly component is worth it; looks simple enough to do them on a case-by-case basis, but the more "machinery" you can remove, so that it's obvious exactly what the plot is doing, the better.

Theodlz commented on skyportal/handlers/api/config_handler.py:
```diff
@@ -96,8 +98,7 @@ def get(self):
-                "colorPalette": getattr(
-                    palettes, cfg.get("misc.bokeh_color_palette", "Turbo256")
-                ),
+                "colorPalette": COLOR_PALETTE,
```
Sounds like a good idea. Originally I felt like not adding more to the config would help with its readability, but it's a nice option to have it be configurable.

Theodlz replied:
it's a very long list however. If I do add it to the yaml config, maybe it should at least be on just one line to not double its current length. I'm not sure this palette is meant to be modified. It's grabbing the nice bokeh color palette we were using to avoid keeping bokeh as a dependency just for that.

Theodlz commented on skyportal/tests/frontend/sources_and_observingruns_etc/test_sources.py:
```diff
@@ -669,20 +631,35 @@ def test_dropdown_facility_change(driver, user, public_source):
-        f'//div[@data-testid="group_{public_group.id}"]',
-        scroll_parent=True,
-    )
+    # popup to close it, and retry a few times in case the page loads slowly
```
So it's not really a hack, it's what really happens in the frontend. You click on a dropdown, and to close the dropdown you just click anywhere outside of it.

Theodlz commented on skyportal/tests/frontend/test_plotly_plots_frontend.py:
```diff
@@ -0,0 +1,37 @@
+
+    button_present = False
+    nretries = 0
+    while button_present is False and nretries < 20:
```
Do you have any suggestions? The idea is that our page has N number of elements that render in different order and speeds based on the availability of the API, the browser used, ... a ton of different factors.

Ideally, we should open a PR and replace all of the things that need to load by skeletons that ensure a fixed size before we load the data. That will allow us not to have these retries to check that something loads.

Because if we trade this pattern for an explicit wait that everything else has loaded, it might be a little tricky as it depends for each test... This is why I had a time.sleep() above, which I traded for a while loop so that if it works on the first try we don't wait for nothing.

Theodlz commented on static/js/components/InstrumentSummary.jsx:
```diff
@@ -96,8 +65,19 @@ const InstrumentSummary = ({ route }) => {
+      if (response.status !== "success") {
+        dispatch(showNotification("Error fetching instrument logs", "error"));
+      }
+      setLoading(false);
```
No it doesn't, that's correct. However I would prefer having a dedicated PR for that. I refactored/wrote some of that code to avoid existing issues but any other improvement would deserve a PR of their own.

stefanv replied:
That works, please file an issue.

Theodlz commented on static/js/components/InstrumentSummary.jsx:
```diff
@@ -171,15 +158,22 @@ const InstrumentSummary = ({ route }) => {
+                  {!loading &&
+                    instrument.log_exists &&
+                    instrument?.logs?.length === 0 && (
+                      <div> No logs exist in the specified time range </div>
```
Same comment as above

Theodlz commented on static/js/components/PhotometryDownload.jsx:
```diff
@@ -0,0 +1,68 @@
+      }
+      return newPoint;
+    });
+    let csvText = `#source: "${obj_id}" downloaded at ${new Date().toISOString()} UTC\nmjd,filter,flux,fluxerr,zp,magsys,lim_mag,stacked\n`;
+    photometryData.forEach((point) => {
+      csvText += `${point.mjd},${point.filter},${point.flux},${point.fluxerr},${point.zp},${point.magsys},${point.lim_mag},${point.stacked}\n`;
+    });
+    const blob = new Blob([csvText], { type: "text/csv;charset=utf-8;" });
```
We do what we can hahaha

Theodlz commented on static/js/components/PhotometryPlot.jsx:
```diff
@@ -0,0 +1,1336 @@
+        return names.length === 0;
+      });
+      newPoint.text = `MJD: ${newPoint.mjd.toFixed(6)}
+        `;
```
Nice catch, thanks!

Theodlz commented on static/js/components/PhotometryPlot.jsx:
```diff
@@ -0,0 +1,1336 @@
+        newPoint.text += `<br>Origin: ${newPoint.origin}`;
+      }
+      if (
+        newPoint.altdata?.exposure !== null &&
```
We could... Yeah having a nice methods that does:
```js
const empty = (value) => {
    return ([null, undefined, "", "None", "undefined"].includes(value) === false)
}
```
could be very useful.

That might require a very large scale refactoring to apply it everywhere we need it very thoroughly, so maybe we can open an issue and have a dedicated PR?

stefanv replied:
Yes, let's add that issue—thanks!

Theodlz commented on static/js/components/PhotometryPlot.jsx:
```diff
@@ -0,0 +1,1336 @@
+          const colorRGB = filter2color[groupedPhotometry[key][0].filter] || [
+            0, 0, 0,
+          ];
+          const colorBorder = `rgba(${colorRGB[0]},${colorRGB[1]},${colorRGB[2]}, 1)`;
```
Will do!

Theodlz commented on static/js/components/SourceDesktop.jsx:
```diff
@@ -698,21 +699,35 @@ const SourceDesktop = ({ source }) => {
               <PhotometryToolTip />
             </AccordionSummary>
-            <AccordionDetails>
+            <AccordionDetails style={{ paddingTop: 0, marginTop: 0 }}>
```
We mostly do not. However from time to time when the styling isn't reused anywhere on the page it's simpler to use that than the classes.<classname> notation. It's easier in terms of readability.

BUT, here we do reuse it so I'll move it in the `classes`

Theodlz commented on static/js/components/SourceDesktop.jsx:
```diff
@@ -698,21 +699,35 @@ const SourceDesktop = ({ source }) => {
+                        annotations={source?.annotations || []}
+                        spectra={spectra || []}
+                        gcn_events={source.gcn_crossmatch || []}
+                        plotStyle={{
```
well we could, it's just that here we customize it based on `rightPanelVisible` which is not a css variable that we can pass to the MUI useStyle method that we rely on. 

We are planning on redesigning the source page and I hope that we will drop this right panel thing or modify it, but until then not sure how to best handle that.

Theodlz commented on static/js/components/SourceMobile.jsx:
```diff
@@ -133,11 +130,12 @@ export const useSourceStyles = makeStyles((theme) => ({
-    flexDirection: "column",
-    paddingBottom: "0.5rem",
-    overflowX: "scroll",
+  plotContainer: {
```
What do you mean?

Theodlz commented on skyportal/handlers/api/photometry.py:
```diff
@@ -54,6 +56,82 @@
+cmap_ir = cm.get_cmap('autumn')
+
+
+def hex2rgb(hex):
```
I think that matplotlib's function expects matplotlib colors, not hex strings. I'll take a look

Theodlz commented on static/js/components/SourceMobile.jsx:
```diff
@@ -719,21 +711,33 @@ const SourceMobile = WidthProvider(
-                      <div> No photometry exists </div>
-                    ) : (
+              <AccordionDetails style={{ paddingTop: 0, marginTop: 0 }}>
+                <Grid container id="photometry-container">
```
Yes, it does work! Not necessary here, but in the future, I would love to design some robust tests that verify that the information that's on the plots is correct, and having that idea to understand what plot we are looking at would be helpful.

Theodlz commented on static/js/components/SpectraPlot.jsx:
```diff
@@ -0,0 +1,974 @@
+  },
+];
+
+function median(values) {
```
At least I could recreate a utils.js file or smth where I move over all of these pure js functions

Theodlz replied:
I did just that. Renamed the existing `plotjs` folder to `utils` and added 3 separate files to it.:
- the constants, things that are static and that we reuse in multiple places.
- the common calculations we might need in multiple places
- calculations specifically dedicated to sky positions

These 3 files could easily be refactored and renamed, but that's a good start.

Theodlz commented:
@stefanv as discussed (at least for the new plots), I moved all of the `style={}` in the usual MUI `makeStyle` that we have on top of each file. I did leave some of them that were so short and not repeated that they didn't seem worth moving somewhere else.

Also, I made sure that editing the photometry from the front end triggers a refresh of the data to keep the plots up to date.

(also removed a langchain warning, just mentioning it here so we know why I'm adding an unrelated dependency in this PR)
